### PR TITLE
Forbids use of `_` anonymous variable

### LINKED ISSUE #309: Limit how `_` variable can be used
# Rule request

## Thesis

Today I saw this code:

```python
[redis_db.delete(symbol_lock_key(_, endpoint_name)) for _ in symbols if True or redis_db.get(symbol_lock_key(_, endpoint_name)) == upid]
```

The thing about it is that we are using `_` as a regular variable.
We also need to list this special case in our `naming.rst`.

## Reasoning

`_` is a special case that indicates that this variable is not used.
So, when it is used we need to alert the user.

### PR DESCRIPTION
**Pull request by nixphix**
Closes #309

Added rule for `_` variable name **Z119**

## Checklist

- [x] I have double checked that there are no unrelated changes in this pull request (old patches, accidental config files, etc)
- [x] I have created at least one test case for the changes I have made
- [x] I have updated the documentation for the changes I have made
- [x] I have added my changes to the `CHANGELOG.md`

Adding `_` check raised violation for existing code in the repo, need to fix them.

### CODE REVIEW

File: wemake_python_styleguide/logics/naming/logical.py [hunk 1]
```diff
@@ -160,3 +160,35 @@ def does_contain_consecutive_underscores(name: str) -> bool:
...
+    False
+
+    >>> is_anonymous_variable('_')
+    True
+
+    """
+    if name == '_':
+        return True
+
+    return False
```
Comment (42B) [inline wemake_python_styleguide/logics/naming/logical.py]:
```python
return name == '_'
```

File: wemake_python_styleguide/violations/naming.py [hunk 2]
```diff
@@ -519,3 +521,36 @@ class ConsecutiveUnderscoresInNameViolation(MaybeASTViolation):
...
+    .. versionadded:: 0.5.0
+
+    Note:
+        Returns Z119 as error code
+
+    """
+
+    #: Error message shown to the user.
+    error_template = 'Found anonymous variable: {0}'
+    code = 119
```
Comment (sobolevn) [inline wemake_python_styleguide/violations/naming.py]: Why is it `119` and not `117`?

File: wemake_python_styleguide/visitors/ast/naming.py [hunk 3]
```diff
@@ -60,11 +61,17 @@ def _check_name(self, node: ast.AST, name: str) -> None:
 
         if logical.does_contain_underscored_number(name):
             self.add_violation(UnderscoredNumberNameViolation(node, text=name))
+
         if logical.does_contain_consecutive_underscores(name):
             self.add_violation(
                 ConsecutiveUnderscoresInNameViolation(node, text=name),
             )
 
+        if logical.is_anonymous_variable(name):
```
Comment (sobolevn) [inline wemake_python_styleguide/visitors/ast/naming.py]: That's not how this should work.
We should forbid usage of the variable name, not its declaration. What do I mean by usage:
- `print(_)`
- `4 * _`
- `x = _`

What do I mean by declaration?
- `some, _ = call()`
- `(do_some() for _ in range(5))`

Comment (nixphix) [reply wemake_python_styleguide/visitors/ast/naming.py]: I'm going through `green tree snakes` document to implement this rule. Meanwhile, let me know if there is a similar implementation in the repo.

Comment (sobolevn) [reply wemake_python_styleguide/visitors/ast/naming.py]: Nope, I do not think so. You need to get the context of `ast.Name`. It should not be `Load` for this variable.

**Comment by nixphix** [comment]
for some reason `test_all_visitors_contained_in_checker` is not passing, does this require explicit configuration of the new Node Visitor ?

**Comment by sobolevn** [comment]
@nixphix add your new visitor to the preset.

**Comment by sobolevn** [comment]
@nixphix thanks a lot! Awesome work!

P.S. Do you have any feedback? How was your developer experience? Docs quality? Review process? How did you like the codebase quality and style? Anything, generally!

**Comment by nixphix** [comment]
> @nixphix thanks a lot! Awesome work!
> 
> P.S. Do you have any feedback? How was your developer experience? Docs quality? Review process? How did you like the codebase quality and style? Anything, generally!

It was great, got to know about AST better.  Using `Poetry` for the first time.

**Pull request merged**